File: python/0202快乐数.py

def isHappy(n: int) -> bool:
    memo=set()

    while(True):
        s=sum([int(i)*int(i) for i in str(n)])

        if(s==1):
            return True
        if(s in memo):
            return False

        memo.add(s)
        n=s

File: cpp/0202快乐数.cpp
#include<unordered_set>

using namespace std;


bool isHappy(int n)
{
    unordered_set<int> memo;
    int s;
    while (true)
    {
        if(n==1)
            return true;
        if(memo.find(n)!=memo.end())
            return false;

        memo.insert(n);
        s = 0;
        while (n > 0)
        {
            s += (n % 10) * (n % 10);
            n /= 10;
        }
        n = s;


    }
}